Feature/contact form in ecommerce hubspot integration

mattwendzina opened the pull request:
Closes https://github.com/Open-Study-College/osc/issues/724
Closes https://github.com/Open-Study-College/osc/issues/728
Closes https://github.com/Open-Study-College/osc/issues/729

## 📝 Description

This builds upon the `feature/form-component-validation-updates` & `contact-form-in-ecommerce` branches by adding the ability to dynamically import the form inputs directly from hubspot.  

## ⛳️ Current behavior (updates)

N/A

## 🚀 New behavior

- Adds form component validation for server and client side using Zod
- Adds Hubspot integration:
1) Enables forms to be created and will pull in the form input and data
2) Allows these forms to be submitted back to Hubspot
- Adds Sanity integration:
1) Enables Marketing to select a form Hubspot
2) Adds settings to enable the form to be a slide-out form if required

## 💣 Is this a breaking change (Yes/No):
No

## 📝 Additional Information

- Currently the formIds are being hardcoded. Eventually want these to be pulled in dynamically using the Hubspot Forms API. This will likely need a proxy and will be re-visited further down the line.

stevenbister commented on packages/osc-ecommerce/app/components/Forms/FormContainer.tsx:
```diff
@@ -0,0 +1,38 @@
+    return (
+        <div className={slideOut ? classes : 'c-form__container'}>
+            {slideOut ? (
+                <Button variant="quaternary" className="c-form__container--slide-out-btn">
```
I think rather than a modifier this would be an element, for example `c-form__slide-out-btn`

stevenbister commented on packages/osc-ecommerce/app/components/Forms/HubspotForm/HubspotForm.spec.tsx:
```diff
@@ -0,0 +1,48 @@
+                setValidationErrors={() => {}}
+                submitText={textContent.submitText}
+                termsAndConditions={textContent.termsAndConditions}
+                titleAndDescription={textContent.titleAndDescription}
```
Looks like a few of these props aren't being used anymore?

mattwendzina replied:
Removed.

stevenbister commented on packages/osc-ecommerce/app/components/Forms/HubspotForm/textContent.ts:
If we update the test file to remove those props will we still need this file?

mattwendzina replied:
Yeah good point, there was still some data being used, but moved it all into the mockData and organised it a bit better so it's clear what mockData is Hubspot and which is Sanity.

stevenbister commented on packages/osc-ecommerce/app/components/Forms/types.ts:
Assuming that these types are pulled from the Hubspot docs? But have a few questions/suggestions
- Worth changing the comments to JSDoc?
- Do we need to include the deprecated ones? If so I think you can mark them as deprecated with JSDoc `@deprecated`
- Worth making the readonly ones readonly with ts? https://www.typescriptlang.org/docs/handbook/utility-types.html#readonlytype

Appreciate we might not be directly hitting these so happy to leave it as is if it's a lot of effort

mattwendzina replied:
I think all the above is good. Done.

stevenbister commented on packages/osc-ecommerce/app/components/Forms/utils.tsx:
```diff
@@ -0,0 +1,113 @@
+                        schema={schema.pick({ [hubspotFields.name]: true })}
+                        setErrors={setValidationErrors}
+                        type={hubspotFields.fieldType}
+                        // variants={data.variants}
```
Should this be commented out?

mattwendzina replied:
I think this was used on some of the other forms, but can't remember now lol. I've removed and will just add back in if needed.

stevenbister commented on packages/osc-ecommerce/app/routes/$slug.tsx:
```diff
@@ -47,9 +66,48 @@ export const loader: LoaderFunction = async ({ request, params }) => {
     });
 };
 
+export const action: ActionFunction = async ({ request }: ActionArgs) => {
```
Rather than having this in here it might be worth moving it into it's own route (`actions/hubspot` or something) and updating the `RemixForm` action to post to that path.

Doing it in the the route means we'll have to duplicate this in each of the other route files (collections, products, posts etc.)

What do you think?

mattwendzina replied:
Yep, thanks mate, this was part of the flow I hadn't fully figured out. Makes sense. Implemented. Took a while though as I hadn't used useFetcher so initially it was reloading the page with Remix Form, and then it took some more time to work out that the data gets returned via useFetcher rather than useLoaderData 😅 
Got there eventually though 😄

stevenbister replied:
Nooiiice

stevenbister commented on packages/osc-ecommerce/app/utils/hubspot.helpers.ts:
```diff
@@ -0,0 +1,98 @@
+    formFieldsData: Record<any, string>[],
+    formData: Record<any, string>
+) => {
+    debugger;
```
Can we take this out?

mattwendzina replied:
Ooops lol. Done.

stevenbister commented on packages/osc-ecommerce/app/utils/hubspot.helpers.ts:
```diff
@@ -0,0 +1,98 @@
+    try {
+        const response = await hubspotFormsApiRequest(
+            'post',
+            `https://api.hsforms.com/submissions/v3/integration/secure/submit/${process.env.CONTACT_PORTAL_ID}/${formId}`,
```
Might be worth adding this env to the README if you've not already

stevenbister commented on packages/osc-ecommerce/app/utils/server/hubspot.server.ts:
```diff
@@ -0,0 +1,48 @@
+
+export const hubspotClient = () => {
+    return new Client({
+        accessToken: process.env.HUBSPOT_DEV_ACCESS_TOKEN,
```
Might be worth adding this env to the README if you've not already

mattwendzina replied:
Added.

stevenbister reviewed (changes requested):
Looking good man! Left a few comments and thoughts. I'm also getting an error when running locally
![image](https://user-images.githubusercontent.com/23461173/226639272-b67f7079-9806-4f19-b108-b0cbf8d0aa34.png)

Think we might have to add any environment variables into GitHub/the workflows as well. Happy to have that in here or as a separate PR, whichever you think is easiest

mattwendzina commented on packages/osc-ecommerce/app/components/Forms/utils.tsx:
```diff
@@ -0,0 +1,113 @@
+        | InputHTMLAttributes<HTMLInputElement>
+        | TextareaHTMLAttributes<HTMLTextAreaElement>;
+
+    // TODO - Need to work out the correct type for this!
```
Updated

stevenbister commented on packages/osc-ecommerce/app/routes/$slug.tsx:
```diff
@@ -37,6 +39,22 @@ export const loader: LoaderFunction = async ({ request, params }) => {
+        };
+    } catch (error) {
+        console.error('Unable to load form!');
+    }
```
This whole block will need adding to all of the routes as we'll want the option to use a form on any of them

mattwendzina replied:
I've refactored into one helper function and added to all same routes where modules are imported.

stevenbister reviewed (changes requested):
Good stuff man, last couple of bits and I think we're there!

Something we'll need to think about doing is adding in success messages once the form submits "Thank you, we'll be in touch soon" kind of stuff, and how we handle that. Probs a separate task, I'll pop a note into Whimsical and perhaps we can discuss in the next sprint planning session?

mattwendzina commented on packages/osc-ecommerce/app/components/Forms/utils.tsx:
```diff
@@ -0,0 +1,112 @@
+    const isSubmitting = fetcher.state === 'submitting';
+    const isAdding =
+        fetcher.state === 'submitting' &&
+        fetcher.submission.formData.get('_action') === 'submitForm';
```
Nice spot! Done.

mattwendzina commented on packages/osc-ecommerce/app/components/Forms/HubspotForm/HubspotForm.tsx:
```diff
@@ -0,0 +1,76 @@
+                        validationErrors
+                    );
+                })}
+                <Button
```
Good idea. Done.

stevenbister reviewed (approved):
Nice man! LGTM 🚀

mattwendzina commented:
@stevenbister - I've had to make some additional changes..! After finishing I went and added a second form to start working on the newsletter to discover that it just displayed the contact form twice 😬 
Realised I was using a `.find()` on module.forms and so it only returned first one. I've changed this to a filter and then had to introduce some additional shaping and logic to ensure that the right form module gets the correct hubspot data.

mattwendzina commented:
Also needed to tweak the ids for the form components to make sure they are unique for the case of multiple forms where there could be overlap.

stevenbister commented:
> @stevenbister - I've had to make some additional changes..! After finishing I went and added a second form to start working on the newsletter to discover that it just displayed the contact form twice 😬 Realised I was using a `.find()` on module.forms and so it only returned first one. I've changed this to a filter and then had to introduce some additional shaping and logic to ensure that the right form module gets the correct hubspot data.

Ahaa oops! Tbf I didn't test putting multiple forms on a page 😅 I'll give it another once over now

stevenbister commented on packages/osc-ecommerce/app/routes/$slug.tsx:
```diff
@@ -3,12 +3,13 @@ import { json } from '@remix-run/node';
 import { useLoaderData, useParams } from '@remix-run/react';
 import { useState } from 'react';
 import type { DynamicLinksFunction } from 'remix-utils';
+import type { HubspotFormData } from '~/components/Forms/types';
```
Eslint warning here `'HubspotFormData' is defined but never used`

mattwendzina replied:
Removed 👍

stevenbister reviewed (changes requested):
Just that eslint warning but otherwise the new changes look good to me
